### PR TITLE
Legend: extract common renderLegendFormat function (loki+prom)

### PR DESCRIPTION
**Pull request by ryantxu**
There are currently 2 almost identical labels template formatter functions -- this moves them to `@grafana/data` so it can be used consistently in StreamingDataFrame.

My regex foo is not strong enough to combat this:
image

### CODE REVIEW

File: packages/grafana-data/src/utils/labels.ts [hunk 1]
```diff
@@ -48,6 +48,12 @@ export function findCommonLabels(labelsSets: Labels[]): Labels {
   }, (undefined as unknown) as Labels);
 }
 
+/** replace labels in a string.  Used for loki+prometheus legend formats */
+export function renderLabelsTemplate(aliasPattern: string, aliasData: Labels): string {
+  const aliasRegex = /\{\{\s*(.+?)\s*\}\}/g;
```
Comment (leeoniya) [inline packages/grafana-data/src/utils/labels.ts:53]: maybe?

```suggestion
const aliasRegex = /\{\{\s*([^{}]+?)\s*\}\}/g;
```

Comment (leeoniya) [reply packages/grafana-data/src/utils/labels.ts:53]: but probably this, which is resilient to bad syntax. this one also assumes there cannot be spaces in the variable name itself.

```suggestion
  const aliasRegex = /(?:\{+\s*)([^{}\s]+)(?:\s*\}+)/g;
```

Comment (ryantxu) [reply packages/grafana-data/src/utils/labels.ts]: Not sure if we need/want to support the these variations, but the pattern above fails for:
image

**Comment by ryantxu** [comment]
@aocenas @torkelo -- thoughts here?

I figured this would be a simple restructure, but looks like our CodeQL settings won't let us merge this regular expression.  Should we change it, or just override the setting?  I am concerned about changing this since it is used so widely already

**Comment by aocenas** [comment]
@ryantxu don't think CodeQL is something that blocks merge. In any case this feels like could be done with some code + some regex maybe? Not sure if speed here is to priority, seems like doing it with some split by or string scan may be more readable anyway.

**Comment by leeoniya** [comment]
if inline regex complexity is too high, you can also assemble it from parts and comment each, though it requires additional escapes for backslashes:

```js
const aliasRegex = new RegExp(
  '(?:\\{+\\s*)' +  // one or more { followed by all trailing whitespace
  '([^{}]+)' +      // capture all non-brace characters
  '(?:\\s*\\}+)'    // all leading whitespace followed by one or more }
);
```

**Comment by ryantxu** [comment]
After messing with the regex enough to think we could make it different, but maybe not *better* I think we should just keep as is ignoreling the CodeQL failure.  This just touches too many things to justify changing it without a really good reason.

**Review by mckn — APPROVED** [review]
The code looks good! 👍🏻 

Where are we planning on using this apart from in prometheus and Loki? I was just thinking if it could live in Grafana core instead of data. But I guess it depends on if we need to use it somewhere else.

**Comment by torkelo** [comment]
Not sure we should use `{{ label }}` syntax more but try to align on the format we use everywhere else, meaning `${variable}`  , best would be to use the same as we have in displayName, ` ${__field.labels.server}` but maybe a bit verbose

**Comment by torkelo** [comment]
@ryantxu where do you need this, anyway you can just use:

```
 templateSrv.replace(alias, { ...request.scopedVars ...labels })  
```

**Comment by ryantxu** [comment]
> where do you need this, anyway you can just use:

For LokiLive -- I need to expand the legendFormat into field names -- and want it to have the same behavior as loki+prometheus do.  (looks similar but not the same as template rendering)

**Comment by ryantxu** [comment]
> Where are we planning on using this apart from in prometheus and Loki? I was just thinking if it could live in Grafana core instead of data. But I guess it depends on if we need to use it somewhere else.

With the change to move StreamingDataFrame to core https://github.com/grafana/grafana/pull/41492  -- I think we can keep this is core now :tada:

**Pull request merged**